File: apps/backends/node/src/db/queries.ts
import { Pool } from 'pg';
import { v4 as uuidv4 } from 'uuid';
import { Product, Order, OrderItem } from '../types.js';

export async function getProduct(pool: Pool, productId: string): Promise<Product> {
  const result = await pool.query(
    'SELECT id, price_cents, stock FROM bakeoff_node.products WHERE id = $1',
    [productId]
  );
  if (result.rows.length === 0) {
    throw new Error('Product not found');
  }
  return result.rows[0];
}

export async function insertOrder(
  pool: Pool,
  order: Order,
  items: OrderItem[]
): Promise<void> {
  const client = await pool.connect();
  try {
    await client.query('BEGIN');

    // Insert order
    await client.query(
      'INSERT INTO bakeoff_node.orders (id, customer_id, total_cents, tax_cents, created_at) VALUES ($1, $2, $3, $4, NOW())',
      [order.id, order.customer_id, order.total_cents, order.tax_cents]
    );

    // Insert items and update stock
    for (const item of items) {
      await client.query(
        'INSERT INTO bakeoff_node.order_items (id, order_id, product_id, quantity, price_cents, created_at) VALUES ($1, $2, $3, $4, $5, NOW())',
        [uuidv4(), order.id, item.product_id, item.quantity, item.price_cents]
      );

      const result = await client.query(
        'UPDATE bakeoff_node.products SET stock = stock - $1 WHERE id = $2 AND stock >= $1',
        [item.quantity, item.product_id]
      );

      if (result.rowCount === 0) {
        throw new Error('Insufficient stock');
      }
    }

    await client.query('COMMIT');
  } catch (error) {
    await client.query('ROLLBACK');
    throw error;
  } finally {
    client.release();
  }
}

File: apps/backends/python/app/db/queries.py
import asyncpg
import uuid as uuid_module
from typing import List

async def get_product(pool: asyncpg.Pool, product_id: str) -> dict:
    row = await pool.fetchrow(
        'SELECT id, price_cents, stock FROM bakeoff_python.products WHERE id = $1',
        uuid_module.UUID(product_id)
    )
    if row is None:
        raise ValueError('Product not found')
    return dict(row)

async def insert_order(
    pool: asyncpg.Pool,
    order_id: str,
    customer_id: str,
    total_cents: int,
    tax_cents: int,
    items: List[dict],
) -> None:
    async with pool.acquire(timeout=5.0) as conn:
        async with conn.transaction():
            # Insert order
            await conn.execute(
                'INSERT INTO bakeoff_python.orders (id, customer_id, total_cents, tax_cents, created_at) VALUES ($1, $2, $3, $4, NOW())',
                uuid_module.UUID(order_id),
                uuid_module.UUID(customer_id),
                total_cents,
                tax_cents,
            )

            # Insert items and update stock
            for item in items:
                await conn.execute(
                    'INSERT INTO bakeoff_python.order_items (id, order_id, product_id, quantity, price_cents, created_at) VALUES ($1, $2, $3, $4, $5, NOW())',
                    uuid_module.uuid4(),
                    uuid_module.UUID(order_id),
                    uuid_module.UUID(item['product_id']),
                    item['quantity'],
                    item['price_cents'],
                )

                result = await conn.execute(
                    'UPDATE bakeoff_python.products SET stock = stock - $1 WHERE id = $2 AND stock >= $1',
                    item['quantity'],
                    uuid_module.UUID(item['product_id']),
                )

                if result == 'UPDATE 0':
                    raise ValueError('Insufficient stock')
